### PR TITLE
devicetree: Add _VARGS variants to _FOREACH_ marcos

### PR DESCRIPTION
**Pull request by arvinf**
`_FOREACH_` macros do not allow the caller to pass additional arguments
to the `fn`. A series of `_VARGS` variants have been added that allow
the caller to pass arbitrary number of arguments to the `fn`:

```
DT_FOREACH_CHILD_VARGS
DT_FOREACH_CHILD_STATUS_OKAY_VARGS
DT_FOREACH_PROP_ELEM_VARGS
DT_INST_FOREACH_CHILD_VARGS
DT_INST_FOREACH_STATUS_OKAY_VARGS
DT_INST_FOREACH_PROP_ELEM_VARGS
```

This is useful in driver code for example where we have multiple instaces of the driver and each instance needs to loop through its children, e.g.
```
#define BAR_CHILD(inst, parent)   \
              .... \
              struct child_data foo_##parent##_bar_##inst##_data;

#define FOO_DRIVER(inst) \
              .... \
              DT_INST_FOREACH_CHILD_VARGS(inst, BAR_CHILD, inst);

DT_INST_FOREACH_STATUS_OKAY(FOO_DRIVER)
```

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

### CODE REVIEW

File: include/devicetree.h [hunk 1]
```diff
@@ -118,7 +118,7 @@
  *   after lowercasing and changing "@" to "_")
  * - and so on for deeper nodes in the desired node's path
  *
- * @param ... lowercase-and-underscores node names along the node's path,
+ * @param ... variable number of arguments to pass to fn
```
Comment (mbolivar-nordic) [inline include/devicetree.h:121]: This looks like a search/replace error; please leave this as-is.

File: include/devicetree.h [hunk 2]
```diff
@@ -2213,6 +2282,23 @@
...
+ *
+ * @param fn Macro to call for each enabled node. Must accept an
+ *           instance number as its only parameter.
+ * @param ... variable number of arguments to pass to fn
+ *
+ * @see DT_INST_FOREACH_STATUS_OKAY
+ */
+#define DT_INST_FOREACH_STATUS_OKAY_VARGS(fn, ...) \
+	COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT),	\
+		    (UTIL_CAT(DT_FOREACH_OKAY_INST_VARGS_,	\
```
Comment (mbolivar-nordic) [inline include/devicetree.h:2298]: Please use DT_CAT, not UTIL_CAT. We try not to use the util macros because their error messages are harder to read.

File: include/devicetree.h [hunk 3]
```diff
@@ -1500,6 +1515,25 @@
...
+ * treated as "okay".
+ *
+ * @param node_id node identifier
+ * @param fn macro to invoke
+ * @param ... variable number of arguments to pass to fn
+ *
+ * @see DT_FOREACH_CHILD_STATUS_OKAY
+ */
+#define DT_FOREACH_CHILD_STATUS_OKAY_VARGS(node_id, fn, ...) \
+	DT_CAT(node_id, _FOREACH_CHILD_STATUS_OKAY_VARGS)(fn, __VA_ARGS__)
```
Comment (mbolivar-nordic) [inline include/devicetree.h:1535]: Please update macros.bnf.

File: include/devicetree.h [hunk 4]
```diff
@@ -1485,6 +1485,21 @@
...
+ * identifier for the child node. The remaining are passed-in by the caller.
+ *
+ * @param node_id node identifier
+ * @param fn macro to invoke
+ * @param ... variable number of arguments to pass to fn
+ *
+ * @see DT_FOREACH_CHILD
+ */
+#define DT_FOREACH_CHILD_VARGS(node_id, fn, ...) \
+	DT_CAT(node_id, _FOREACH_CHILD_VARGS)(fn, __VA_ARGS__)
```
Comment (mbolivar-nordic) [inline include/devicetree.h:1501]: Please update macros.bnf.

File: include/devicetree.h [hunk 5]
```diff
@@ -1547,6 +1581,26 @@
...
+ *
+ * @param node_id node identifier
+ * @param prop lowercase-and-underscores property name
+ * @param fn macro to invoke
+ * @param ... variable number of arguments to pass to fn
+ *
+ * @see DT_FOREACH_PROP_ELEM
+ */
+#define DT_FOREACH_PROP_ELEM_VARGS(node_id, prop, fn, ...)		\
+	DT_CAT4(node_id, _P_, prop, _FOREACH_PROP_ELEM_VARGS)(fn, __VA_ARGS__)
```
Comment (mbolivar-nordic) [inline include/devicetree.h:1602]: Please update the list of property suffixes at the top of the file.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
I like this idea and think the use case is valid.

Some initial requests.

File: include/devicetree.h (hunk 2 above)
Comment (arvinf) [reply include/devicetree.h:2298]: I followed `DT_INST_FOREACH_STATUS_OKAY`. Should I update that as well?

Comment (mbolivar-nordic) [reply include/devicetree.h:2298]: Sure, if you want to tack on another patch with that; thanks.

**Comment by arvinf** [comment]
Changing `DT_UTIL` to `DT_CAT` is causing some errors.

File: include/devicetree.h [hunk 6]
```diff
@@ -45,6 +45,7 @@
  * _ENUM_UPPER_TOKEN: like _ENUM_TOKEN, but uppercased
  * _EXISTS: property is defined
  * _FOREACH_PROP_ELEM: helper for "iterating" over values in the property
+ * _VARGS: foreach functions with variable number of arguments
```
Comment (mbolivar-nordic) [inline include/devicetree.h:48]: This needs to be the complete suffix instead, i.e. `_FOREACH_PROP_ELEM_VARGS`. It's the way we make sure there are no namespace collisions.

Comment (arvinf) [reply include/devicetree.h:48]: Understood. 

Can you please take a look at the build errors regarding DT_CAT below and advise if it would be ok for me to revert the DT_CAT changes, or if there is another way to fix the issue.

Comment (mbolivar-nordic) [reply include/devicetree.h:48]: I think I see what's going on and a fix is out of scope. Let's just use UTIL_CAT for now and I'll try to change it in a follow-up PR at some point.

**Comment by mbolivar-nordic** [comment]
@arvinf did you forget to push something? There are various CI errors that look related to use of DT_CAT instead of going back to UTIL_CAT as discussed, no?

**Comment by arvinf** [comment]
My mistake. I thought you were asking me to keep the `DT_CAT` change as-is.

**Comment by arvinf** [comment]
mbolivar-nordic - updated with changes. I also discovered another bug, which I've added as a second commit here.

**Comment by mbolivar-nordic** [comment]
> I also discovered another bug, which I've added as a second commit here.

Can you please remove this commit from this pull request? It's not related to adding VARGS and I think it deserves a separate discussion.

The stated reason for converting whitespace to underscores in str2ident is:

```
Prior to this having `compatible = "Ublox,BMD 345 EVAL"` would
result in `#define DT_N_INST_0_ublox_bmd 345 eval DT_N` which is
incorrect. The correct output is
`#define DT_N_INST_0_ublox_bmd_345_eval DT_N`
```

However, I searched through the Linux source tree and I was only able to find the following compatibles with spaces in them:

```
"arteris, flexnoc"
"microchip, at24c64d"
"sirf, atlas7-g2d"
"snps, dwc-eth-qos"
```

None of these are covered in the Linux bindings but all of them appear to be typos to my eye.

The DT spec itself is silent about what types of strings belong in compatible properties, but the clear convention is to not have whitespace there, and I don't think it can simply be handled as fixing a bug.

**Comment by mbolivar-nordic** [comment]
> However, I searched through the Linux source tree and I was only able to find the following compatibles with spaces in them:

Hmm, I had an old linux tree. Even those limited examples are all gone now.

- "arteris, flexnoc" and "sirf, atlas7-g2d": looks like platform support was removed in https://github.com/torvalds/linux/commit/f3a732843accb04ede91055ffd0b92464fa4d15c
- "microchip, at24c64d" was fixed in https://github.com/torvalds/linux/commit/f3a732843accb04ede91055ffd0b92464fa4d15c

The commit message for the microchip case led me to here:

**Comment by arvinf** [comment]
The reason I kept it with this was because my patch would fail without this fix. I discovered the bug because a random test failed with that message. I can take the patch out but the regression test will fail.

The problem shows up because we use the result of `str2ident` to generate the macro name which should not have a space. I don't know if Linux does the same with it but in any event I don't see a harm in ensuring output of str2ident doesn't have a space because of how we use it.

I am also fine with fixing the DTs instead and removing the space, but don't want to leave it at just a fix. We should have a way to enforce the pattern during parsing and generate an error for a user that has a space in the compatible string.

**Comment by arvinf** [comment]
The reason I added that commit to this patch was because my patch would fail without this fix. I discovered the bug because a random regression test failed with an error that was about an incorrect macro name. I can take the commit out but the regression test will fail.

The problem shows up because we use the result of `str2ident` to generate the macro name which should not have a space. I don't know if Linux uses the output same way. I don't see a harm in ensuring output of `str2ident` doesn't have a space, because of how we use it.

I am also fine with fixing the DTs instead as well, but don't want to leave it at just a fix. We should have a way to enforce the pattern during parsing and generate an error for a user that has a space in the compatible string.

So I see three things that need to happen in the fix: 
1. Keep `str2ident` as-is to ensure the output doesn't have a space.
2. Add pattern check to ensure the compatible property conforms to the dtspec and generate an error for the user if it doesn't
3. Fix the violating DTs.

**Comment by mbolivar-nordic** [comment]
>  We should have a way to enforce the pattern during parsing and generate an error for a user that has a space in the compatible string.

Yep, I sent a PR: https://github.com/zephyrproject-rtos/zephyr/pull/35313

**Comment by arvinf** [comment]
I have a PR for this as well. I am generalizing it, more like the Linux approach, and adding a pattern property that can be used to check against any string or string-array. I can push it here for you to review.

**Comment by arvinf** [comment]
Ok. I'll rebase once that PR is merged.

**Comment by arvinf** [comment]
@mbolivar-nordic - rebased to use your PR.

File: tests/lib/devicetree/api/src/main.c [hunk 7]
```diff
@@ -1376,8 +1404,17 @@ static void test_devices(void)
 	 */
 	val = 0;
 #define INC(inst_ignored) do { val++; } while (0);
-	DT_INST_FOREACH_STATUS_OKAY(INC)
+	DT_INST_FOREACH_STATUS_OKAY(INC);
```
Comment (mbolivar-nordic) [inline tests/lib/devicetree/api/src/main.c:1407]: Why? This change doesn't seem related to this PR and it will lead to an extraneous semicolon in the generated output. I'd be nervous some linter somewhere will complain.

File: tests/lib/devicetree/api/src/main.c [hunk 8]
```diff
@@ -1376,8 +1404,17 @@ static void test_devices(void)
...
 	zassert_equal(val, 2, "");
+#undef INC
+
+	val = 0;
+#define INC_ARG(arg) do { val++; val += arg; } while (0)
+#define INC(inst_ignored, arg) INC_ARG(arg);
+	DT_INST_FOREACH_STATUS_OKAY_VARGS(INC, 1);
```
Comment (mbolivar-nordic) [inline tests/lib/devicetree/api/src/main.c:1414]: Same nervousness here.

File: tests/lib/devicetree/api/src/main.c [hunk 9]
```diff
@@ -1387,6 +1424,13 @@ static void test_devices(void)
 #define DT_DRV_COMPAT xxxx
 #define BUILD_BUG_ON_EXPANSION (there is a bug in devicetree.h)
 	DT_INST_FOREACH_STATUS_OKAY(BUILD_BUG_ON_EXPANSION)
+#undef BUILD_BUG_ON_EXPANSION
+
+#undef DT_DRV_COMPAT
+#define DT_DRV_COMPAT xxxx
```
Comment (mbolivar-nordic) [inline tests/lib/devicetree/api/src/main.c:1430]: Why undefine it and define it to the same thing again?

File: tests/lib/devicetree/api/src/main.c [hunk 10]
```diff
@@ -1591,6 +1635,48 @@ static void test_child_nodes_list(void)
...
+
+	struct vnd_child_binding vals_status_okay[] = {
+		DT_FOREACH_CHILD_STATUS_OKAY_VARGS(TEST_PARENT, TEST_FUNC, 1)
+	};
+
+	zassert_equal(ARRAY_SIZE(vals), 3, "");
+	zassert_equal(ARRAY_SIZE(vals_inst), 3, "");
+	zassert_equal(ARRAY_SIZE(vals_status_okay), 2, "");
+
+	zassert_false(strlen(STRINGIFY(TEST_PARENT)) == 0, "");
```
Comment (mbolivar-nordic) [inline tests/lib/devicetree/api/src/main.c:1665]: This assertion does not seem related to the test case. Is it debugging code? Can't it be removed?

**Review by mbolivar-nordic — COMMENTED** [review]
Thanks for rebasing. I have a few minor questions about the test cases but otherwise I think this PR is in great shape.

File: tests/lib/devicetree/api/src/main.c (hunk 9 above)
Comment (arvinf) [reply tests/lib/devicetree/api/src/main.c:1430]: I wanted to keep the test for each feature clean and separate.

File: tests/lib/devicetree/api/src/main.c (hunk 10 above)
Comment (arvinf) [reply tests/lib/devicetree/api/src/main.c:1665]: I followed the code pattern in `test_child_nodes_list`.

Comment (mbolivar-nordic) [reply tests/lib/devicetree/api/src/main.c:1665]: It doesn't look reasonable for it to be in that function either; I've just sent a PR removing it. https://github.com/zephyrproject-rtos/zephyr/pull/36109. Thanks for pointing that out.

File: tests/lib/devicetree/api/src/main.c (hunk 9 above)
Comment (mbolivar-nordic) [reply tests/lib/devicetree/api/src/main.c:1430]: They are so close to each other that I think it's probably more confusing than just leaving it as-is, but okay, I accept that.

File: tests/lib/devicetree/api/src/main.c (hunk 10 above)
Comment (arvinf) [reply tests/lib/devicetree/api/src/main.c:1665]: I'll remove it from my code then and push the changes.

**Comment by mbolivar-nordic** [comment]
Thanks! The failing 'Devicetree script tests' check appears unrelated to your PR. I've sent what I believe is a fix here: https://github.com/zephyrproject-rtos/zephyr/pull/36110

This appears to be due to a recent change to mypy which is affecting our CI.

**Comment by mbolivar-nordic** [comment]
Let's see if we can get my fix to CI all green and merged and then rebase your PR on top, just to be safe, since it's touching the DT scripts. Sorry about that and thanks for your patience. We'll get this PR in soon I hope.

**Comment by arvinf** [comment]
No problem. I appreciate it that you're always very responsive on PRs.

**Comment by mbolivar-nordic** [comment]
Fix is merged, please rebase

**Comment by arvinf** [comment]
@mbolivar-nordic - CI checks have passed

**Review by galak — COMMENTED** [review]
Do we need the non-varg versions?  Can we just have DT_FOREACH_CHILD always take a varg?

**Comment by mbolivar-nordic** [comment]
I'd rather keep them separate if that's OK with you, @galak. The non-vargs versions are easier to use.

**Comment by galak** [comment]
> I'd rather keep them separate if that's OK with you, @galak. The non-vargs versions are easier to use.

By easier do you mean the error message or some actual usage ease?

**Comment by mbolivar-nordic** [comment]
Both. Examples follow.

---------------

1. Errors are more confusing. For example, with this:

```diff
diff --git a/tests/lib/devicetree/api/src/main.c b/tests/lib/devicetree/api/src/main.c
index 209fb2afa5..4b7efec6ef 100644
--- a/tests/lib/devicetree/api/src/main.c
+++ b/tests/lib/devicetree/api/src/main.c
@@ -1286,7 +1286,7 @@ static void test_foreach_prop_elem_varg(void)
        ((2 * DT_PROP_BY_IDX(node_id, prop, idx)) + arg),
 
        int array[] = {
-               DT_FOREACH_PROP_ELEM_VARGS(TEST_ARRAYS, a, TIMES_TWO_ADD, 3)
+               DT_FOREACH_PROP_ELEM_VARGS(TEST_ARRAYS, a, TIMES_TWO_ADD)
        };
 
        zassert_equal(ARRAY_SIZE(array), 3, "");
```


You will get:

```
.../tests/lib/devicetree/api/src/main.c:1289:17: error: expected expression before ')' token
 1289 |   DT_FOREACH_PROP_ELEM_VARGS(TEST_ARRAYS, a, TIMES_TWO_ADD)
      |                 ^
```

Wat.

---------

2. They aren't a drop-in replacement for existing use cases. For example, this doesn't build:

```diff
diff --git a/tests/lib/devicetree/api/src/main.c b/tests/lib/devicetree/api/src/main.c
index 209fb2afa5..d7f2f402fc 100644
--- a/tests/lib/devicetree/api/src/main.c
+++ b/tests/lib/devicetree/api/src/main.c
@@ -1282,11 +1282,11 @@ static void test_foreach_prop_elem(void)
 
 static void test_foreach_prop_elem_varg(void)
 {
-#define TIMES_TWO_ADD(node_id, prop, idx, arg) \
-       ((2 * DT_PROP_BY_IDX(node_id, prop, idx)) + arg),
+#define TIMES_TWO(node_id, prop, idx) \
+       (2 * DT_PROP_BY_IDX(node_id, prop, idx)),
 
        int array[] = {
-               DT_FOREACH_PROP_ELEM_VARGS(TEST_ARRAYS, a, TIMES_TWO_ADD, 3)
+               DT_FOREACH_PROP_ELEM_VARGS(TEST_ARRAYS, a, TIMES_TWO)
        };
 
        zassert_equal(ARRAY_SIZE(array), 3, "");
@@ -1298,7 +1298,7 @@ static void test_foreach_prop_elem_varg(void)
 #define DT_DRV_COMPAT vnd_array_holder
 
        int inst_array[] = {
-               DT_INST_FOREACH_PROP_ELEM_VARGS(0, a, TIMES_TWO_ADD, 3)
+               DT_INST_FOREACH_PROP_ELEM_VARGS(0, a, TIMES_TWO)
        };
 
        zassert_equal(ARRAY_SIZE(inst_array), ARRAY_SIZE(array), "");
```

I get `error: macro "TIMES_TWO_ADD" passed 4 arguments, but takes just 3`, and the fix is not obvious:

```c
#define TIMES_TWO(node_id, prop, idx, ignored)			\
   ...
```

I have to manually ignore an empty argument in my `fn` macro.

**Review by mbolivar-nordic — APPROVED** [review]
Sorry, I totally forgot I had a -1 on this and not a +1.

LGTM, thanks for all the work. @galak can you please see my above comment on why the separate VARGS versions make sense to me and merge if OK?

**Comment by arvinf** [comment]
@galak  - Any update on this?

**Review by cfriedt — APPROVED** [review]
LGTM.
